### ISSUE TITLE
Unclear docs regarding encryption

### ISSUE THREAD

**Issue by slavko-lukic** [mrousavy/react-native-mmkv#595]
![Screenshot 2023-10-31 at 13 57 36](https://github.com/mrousavy/react-native-mmkv/assets/87854709/19a42b62-2a16-451e-a38c-9693e1b0a0cc)

Based on this sentence in the docs, it is not clear to me if data I save with MMKV is encrypted and safe or not.

Is data encrypted by default or do I need to set encryptionKey explicitly in order for it to be encrypted?
What's the point of setting `encryptionKey` through props when app could be reverse engineered and key would be compromised?

**Comment by jeremiebardon** [comment]
+1 we actually have the same concern in my team. 

How to encrypt and secure the storage to be able to put sensitive information in it, should we ask a token to our Backend and make a rotation of this encryption key or is it already handled by MMKV ?

**Comment by mrousavy** [comment]
100% agree with @francesco-clementi-92 here!

**Comment by NilsBaumgartner1994** [comment]
Is there any way to create and save a key locally in order to use the encryption feature safely?

**Comment by foureight84** [comment]
> 100% agree with @francesco-clementi-92 here!

Definitely agree. However, if a key is lost then is there a way to clear out the encrypted instance? Since you won't be able to create a new instance with the same id and different encryption key. This is my understanding from reading through an older [issue post](https://github.com/mrousavy/react-native-mmkv/issues/409#issuecomment-1664635352) but I want to verify.

**Comment by theabhishek2511** [comment]
> Is there any way to create and save a key locally in order to use the encryption feature safely?

@NilsBaumgartner1994 generate a key separately using [a key derivation function](https://en.wikipedia.org/wiki/Key_derivation_function#Password_hashing) based on some randomness and some user dependent parameters (like their uid), encrypt mmkv using said key, then save it securely.

some libs are available for kdfs, like [react-native-argon2](https://github.com/poowf/react-native-argon2).

to save the key use either [android's keystore](https://developer.android.com/privacy-and-security/keystore) or [ios' keychain](https://developer.apple.com/documentation/security/keychain_services), or just use [react-native-keychain](https://github.com/oblador/react-native-keychain) or [expo securestore](https://docs.expo.dev/versions/latest/sdk/securestore/) if you're on expo.

**Comment by slavko-lukic** [comment]
> > Is there any way to create and save a key locally in order to use the encryption feature safely?
> 
> @NilsBaumgartner1994 generate a key separately using [a key derivation function](https://en.wikipedia.org/wiki/Key_derivation_function#Password_hashing) based on some randomness and some user dependent parameters (like their uid), encrypt mmkv using said key, then save it securely.
> 
> some libs are available for kdfs, like [react-native-argon2](https://github.com/poowf/react-native-argon2).
> 
> to save the key use either [android's keystore](https://developer.android.com/privacy-and-security/keystore) or [ios' keychain](https://developer.apple.com/documentation/security/keychain_services), or just use [react-native-keychain](https://github.com/oblador/react-native-keychain) or [expo securestore](https://docs.expo.dev/versions/latest/sdk/securestore/) if you're on expo.

This kinda defeats the purpose of secure mmkv if you are already using keychain, than you can just use keychain.
One good way I found is combination of keychain and zustand where keychain is used as persist storage.
In that case, data is stored securely in the keychain, you decrypt it once at app start when you hydrade the zustand store, and than you can read the data from memory without the need to decrypt it every time.

**Comment by NilsBaumgartner1994** [comment]
> > This kinda defeats the purpose of secure mmkv if you are already using keychain, than you can just use keychain.
> 
> I don't think it's a good idea to use keychain to store large data (consider the name referring to a store/holder for keys). ios' security daemon [doesn't like having to work with data more than 4kb](https://github.com/square/Valet/issues/246#issuecomment-691793972), for example.
> 
> imho it would still be more robust to use mmkv storage to store your actual encrypted data, alongside keychain to store the key you encrypted mmkv with. especially if you are generating keys locally and not deriving them from a server token or whatever.

True, even considering saving something in web would make the process easier.

but what I would like to wish would be a in app written key saver. So we don’t have to reimplement the same key saving mechanic for android, web and iOS.
Would be easier for many to use mmkv

**Comment by ericlewis** [comment]
> > This kinda defeats the purpose of secure mmkv if you are already using keychain, than you can just use keychain.
> 
> I don't think it's a good idea to use keychain to store large data (consider the name referring to a store/holder for keys). ios' security daemon [doesn't like having to work with data more than 4kb](https://github.com/square/Valet/issues/246#issuecomment-691793972), for example.
> 
> imho it would still be more robust to use mmkv storage to store your actual encrypted data, alongside keychain to store the key you encrypted mmkv with. especially if you are generating keys locally and not deriving them from a server token or whatever.

I was curious about this, since it's such an old comment you are referencing, and I can reliably save 100 megabytes of data in keychain. Using the same testing method. Not sure if device dependent, but likely OS dependent.

**Comment by NilsBaumgartner1994** [comment]
> > > This kinda defeats the purpose of secure mmkv if you are already using keychain, than you can just use keychain.
> > 
> > 
> > I don't think it's a good idea to use keychain to store large data (consider the name referring to a store/holder for keys). ios' security daemon [doesn't like having to work with data more than 4kb](https://github.com/square/Valet/issues/246#issuecomment-691793972), for example.
> > imho it would still be more robust to use mmkv storage to store your actual encrypted data, alongside keychain to store the key you encrypted mmkv with. especially if you are generating keys locally and not deriving them from a server token or whatever.
> 
> I was curious about this, since it's such an old comment you are referencing, and I can reliably save 100 megabytes of data in keychain. Using the same testing method. Not sure if device dependent, but likely OS dependent.

https://docs.expo.dev/versions/latest/sdk/securestore/

„ Size limit for a value is 2048 bytes. An attempt to store larger values may fail.“

**Comment by NilsBaumgartner1994** [comment]
So this is still an issue right?

**Comment by christo8989** [comment]
> > > This kinda defeats the purpose of secure mmkv if you are already using keychain, than you can just use keychain.
> > 
> > 
> > I don't think it's a good idea to use keychain to store large data (consider the name referring to a store/holder for keys). ios' security daemon [doesn't like having to work with data more than 4kb](https://github.com/square/Valet/issues/246#issuecomment-691793972), for example.
> > imho it would still be more robust to use mmkv storage to store your actual encrypted data, alongside keychain to store the key you encrypted mmkv with. especially if you are generating keys locally and not deriving them from a server token or whatever.
> 
> True, even considering saving something in web would make the process easier.
> 
> but what I would like to wish would be a in app written key saver. So we don’t have to reimplement the same key saving mechanic for android, web and iOS. Would be easier for many to use mmkv

Does anyone know if this exists? Seems like a common use case someone might've solved in another package.

**Comment by mrousavy** [comment]
Closing since there is nothing I can do about this in react-native-mmkv.

**Comment by Andrew-Chen-Wang** [comment]
This was closed 9 hours ago which I find funnily coincidental, but I was scouring/looking for what others were doing since it seemed really counterintuitive to hard code your encryption key. Twas right in that regard, but it was misleading in the docs where I initially assumed hard coding the key was fine.

At the very least, referencing this issue #595 in the README would be great. Specifically, pointing people in some direction like https://github.com/mrousavy/react-native-mmkv/issues/595#issuecomment-2026859426 i.e. a react native keychain or sending the encryption key from your backend server
